### ISSUE TITLE
Split `Callable.bind()` into `bind_left` and `bind_right`

### ISSUE THREAD

**Issue by nathanfranke** [godotengine/godot-proposals#4920]
### Describe the project you are working on

https://github.com/godotengine/godot-proposals/discussions/4878

### Describe the problem or limitation you are having in your project

Following the above discussion, looks like multiple people have use cases for binding to the left.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Rename `bind` to `bind_right` to preserve feature compatibility
Add `bind_left` that appends bind arguments to the left of passed arguments.

Internally, this means adding another list since binds are stored before the actual parameters are known.
https://github.com/godotengine/godot/blob/master/core/variant/callable_bind.cpp#L91-L99

**It will probably make sense to do the same changes with `unbind`- `unbind_left` and `unbind_right`.**

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```gdscript
func test(a, b, c):
	pass

func _ready():
	test.bind_left("a").bind_right("c").call("b") # a, b, c passed
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Cannot be worked around without creating a new method or lambda.

### Is there a reason why this should be core and not an add-on in the asset library?

Callables are core.

**Comment by Mickeon** [comment]
One minor concern I personally have is that these would a bit more laborious to read back when reviewing code...

**Comment by nathanfranke** [comment]
Should this dialog be changed?
![image](https://user-images.githubusercontent.com/14253836/179832468-10c71be8-d434-450f-9346-0fd745a8afd6.png)

> One minor concern I personally have is that these would a bit more laborious to read back when reviewing code...

This is similar to the `push_back`, `push_front`, and `append` question. Maybe we should have a sane default and another method/argument for the other way.

**Comment by Mickeon** [comment]
> Although left/right or front/back isn't too much code given that it's more readable.

That is true. Indeed I'd normally prioritise clarity, too, but the example in the proposal, as well as the general assumptions, show the best case scenario for this, that being very short constant Strings passed as arguments. With actual, readable variable names (which Godot really encourages), perhaps not with one parameter, but by the second it may really add up, even if admittedly one could just split the call in multiple lines to make it considerably easier to analyse.

```GDScript
func _ready():
    var properly_named_variable = 0
    var very_cool_name
    var bubble_count

    test.bind_left(properly_named_variable).bind_right(very_cool_name).call(bubble_count)
    
    test.bind_left(properly_named_variable
    ).bind_right(very_cool_name
    ).call(bubble_count)
```

**Comment by nathanfranke** [comment]
> ```gdscript
> func _ready():
>     var properly_named_variable = 0
>     var very_cool_name
>     var bubble_count
> 
>     test.bind_left(properly_named_variable).bind_right(very_cool_name).call(bubble_count)
>     
>     test.bind_left(properly_named_variable
>     ).bind_right(very_cool_name
>     ).call(bubble_count)
> ```

In practice, this code will never be this complicated ass you can just use `test.call(properly_named_variable, bubble_count, very_cool_name)`. I think the most complicated workflow would be `bind_left().bind_right()` in a signal callback.

**Comment by YuriSizov** [comment]
The linked discussion doesn't provide any insight into why it is a limitation. It just states that this is indeed the case that the bound arguments go last.

What you propose introduces a lot of confusion for binding, especially because calls can be chained, in arbitrary order, with multiple additions to either side. Not to mention that a simple and straightforward bind method is turned into a choice paralysis inducing pair of methods.

**Comment by nathanfranke** [comment]
[dalexeev's reply](https://github.com/godotengine/godot-proposals/discussions/4878#discussioncomment-3165258) perfectly demonstrates the limitation. Instead of binding `gradient.set_offset`, they have to declare a lambda that swaps the values.

Desired:
```gdscript
    tween.tween_method(gradient.set_offset.bind_front(1), 0.0, 0.5, 0.8)
    tween.tween_method(gradient.set_offset.bind_front(2), 1.0, 0.5, 0.8)
```

Workaround:
```gdscript
    var set_offset := func (offset, point): gradient.set_offset(point, offset)
    tween.tween_method(set_offset.bind(1), 0.0, 0.5, 0.8)
    tween.tween_method(set_offset.bind(2), 1.0, 0.5, 0.8)
```

This will be even worse if you have to later disconnect the signal (which I needed to). You need to keep a reference to the lambda in a variable or dictionary.

**Comment by KoBeWi** [comment]
I actually had similar idea, but instead of bind_left/right, I'd just do `prebind()`. My use-case was that tweening some methods/callbacks required additional arguments at front. I didn't open a proposal, because it's easy to workaround with lambdas.

> You need to keep a reference to the lambda in a variable or dictionary.

Not really. You can use `get_connections()` in Signal to retrieve the callable.

**Comment by nathanfranke** [comment]
> Not really. You can use `get_connections()` in Signal to retrieve the callable.

This is a pretty messy solution. This would break if another lambda is added.

**Comment by KoBeWi** [comment]
The lambda can be named though if you need to retrieve it (but not sure if the name is accessible).

**Comment by nathanfranke** [comment]
> but not sure if the name is accessible

Pretty sure it's not:
```cpp
StringName CallableCustom::get_method() const {
	ERR_FAIL_V_MSG(StringName(), vformat("Can't get method on CallableCustom \"%s\".", get_as_text()));
}
```

But regardless, still a bit messy :slightly_smiling_face:

**Comment by YuriSizov** [comment]
> I fundamentally disagree. If the user doesn't know whether to use bind_left or bind_right, they shouldn't blindly pick one.

That's not how choice paralysis manifests. Currently you don't need to pick, you have only one option, clearly named, and having a predictable result (once you learn about it). You propose we add two equally prominent options instead, so every time you want to use it, you have to make a decision. Sure, if you're a power user and you really want this flexibility, you'll just enjoy the options. But if you don't actually need this flexibility, you can get stuck deciding what you want to use here, because neither option is really better than the other one.

They will likely end up with a mix of either method without much consistency.

Which is why I like KoBeWi's idea of `prebind` better, because it lets power users to do their thing without making the default API more confusing.

**Comment by nathanfranke** [comment]
What should be the compliment of `unbind`, if any?

**Comment by mieldepoche** [comment]
If `bind` must be kept I'd advocate for `bind_left` and `unbind_left`.
Or similar to String's `split` and `rsplit`: have `bind` and `lbind`.
I don't like "pre" but maybe that's up to personnal preference.

**Comment by Spartan322** [comment]
Given this aims to add to bind anyway, how much complexity would it be to just instead implement a position specifier for bind? Like with another bind method that you send something akin to a dictionary with integer value keys and binds it to an arbitrary position. If the complexity changes for that are relatively low compared to this change you could still perform this and get extra utility out of it, it be somewhat akin to some common use cases of `apply` in Javascript at that point as well. Though perhaps the use case in GDScript is limited and too complex anyhow, I could see it potentially being useful for a lot of utility functionality.

**Comment by samsface** [comment]
boost::bind had a neat interface that solves this.

`bind("known_arg", _1, "known_arg_2", _3)`

**Comment by 3lis4** [comment]
Hi, I think adding a right_bind() / back_bind() (you name it) is not just a cosmetic matter, but is also adding functionality. As of Godot 4.1, bind() always add the argument to the left and after the arguments given to Callable.call(). So it's not possible to my knowledge to do the following in a function :
1) Get a Callable as input, which has some bound arguments
2) Add arguments to the right of its arguments list
3) Call the callable

**Comment by OS-of-S** [comment]
I just wanna add some thoughts into conversation.

I'm not disappointed in the absence of bind_right() or prebind() functions as it is, but I just expected an opposite behavour of the existing bind() function. I mean, that it seems more intuitive and readable to add arguments from left to right.
I also expected that bind() should be applicable for binding an optionall arguments. For exemple, if we pass a Callable with necessary arguments binded and checking in another Object/method can we bind an optional or not. Anyway — we can't. Becouse binding works from left to right.

Moreover, I suppose I have an explanation for a one-way binding. And it's not a style or readability-based as [YuriSizov](https://github.com/YuriSizov) supposed.

My hypothesis that call() functionality under the hood is based on bind(), and bind() is working with Arrays (note that we have [bindv() function](https://docs.godotengine.org/en/stable/classes/class_callable.html#class-callable-method-bindv)). For the performance reason, binding and unbinding (pushing and poping, if we suspect that it's actually an array operations) executed only in one-way order, from the front:

![Image](https://github.com/user-attachments/assets/322f849e-03a2-4338-9e07-bf292d2d9b4c)

IT'S ONLY A HYPOTHESIS, I DID NOT LOOKED INTO SORCE CODE! But it seems logical for me. It still doesn't explain why we binding and unbinding from left to right. It could be connected with the one-string expression and multiline expression — maybe the developers decided that they had to work the same way...

**Comment by Lexyth** [comment]
> boost::bind had a neat interface that solves this.
> 
> `bind("known_arg", _2, "known_arg_2", _4)`
> 
> Where _2 & _4 are placeholders for the 2nd arg passed and the 4th arg passed. Op's proposed interface is nice but doesn't count for if we want a mix and match of args (not just left/right).

Placeholders like that have a limitation in that functions can have any number of arguments, so you'd need to provide an arbitrary amount of placeholders. 

Also, `bind` is currently "inserting" arguments, while boost::bind would replace them instead. 
If it's about replacing them, I'd prefer something along the lines of `fix_arg(arg_index: int, value: Variant) -> void` that fixes the nth argument to the given value and `map_arg(target_index: int, source_index) -> void` that copies an argument from the call.

E.g.: `foo.fix_arg(1, "B").fix_arg(2, "C").map_arg(3, 0).call("a", "b", "c", "d", "e") # effective argList = ("a", "B", "C", "a", "e")`

Utilities could be `fix_args(entries: Dictionary[int, Variant])` and `map_args(mappings: Dictionary[int, int]`, used like this: `foo.fix_args({1: "B", 2: "C"}).map_args({3: 0})`

Edit: I think that duplicating arguments and replacing arbitrary arguments easily and without shifting others around is particularly useful.

**Comment by AriaTheCurseling** [comment]
> I actually had similar idea, but instead of bind_left/right, I'd just do `prebind()`. My use-case was that tweening some methods/callbacks required additional arguments at front. I didn't open a proposal, because it's easy to workaround with lambdas.

`bind` and `prebind` are better names, if for no other reason than backwards compatibility

**Comment by AriaTheCurseling** [comment]
> > I fundamentally disagree. If the user doesn't know whether to use bind_left or bind_right, they shouldn't blindly pick one.
> 
> That's not how choice paralysis manifests. Currently you don't need to pick, you have only one option, clearly named, and having a predictable result (once you learn about it). You propose we add two equally prominent options instead, so every time you want to use it, you have to make a decision. Sure, if you're a power user and you really want this flexibility, you'll just enjoy the options. But if you don't actually need this flexibility, you can get stuck deciding what you want to use here, because neither option is really better than the other one.

There will always be one correct option and one wrong option.

By the time you're binding arguments to a callable, the method signature is already determined and that decides which bind you need.

That said, im againts `bind_right` because it causes magration issues, so i prefer `lbind` and `bind` if we're not going with one of the more powerful designs that can remap the arguments fully

**Comment by dalexeev** [comment]
In 4.5 you can use variadic functions to create a utility function for this purpose:

```gdscript
# utils.gd
class_name Utils

static func callable_prebind(callable: Callable, ...bind_args: Array) -> Callable:
    return func (...call_args: Array) -> Variant:
        return callable.callv(bind_args + call_args)
```

```gdscript
# script.gd
func _ready() -> void:
    var f: Callable = Utils.callable_prebind(prints, "Prefix:")
    f.call(1)       # Prefix: 1
    f.call(1, 2)    # Prefix: 1 2
    f.call(1, 2, 3) # Prefix: 1 2 3
```

Regarding adding the method out of the box, while I agree it could be useful, please keep in mind that it affects other APIs as well:

* `Callable.get_bound_arguments()`;
* `Callable.get_bound_arguments_count()`;
* `Callable.get_unbound_arguments_count()`;
* `Callable.unbind()`.

If we add support for binding/unbinding arguments on both sides, it will significantly overcomplicate the API and make it more confusing.

Currently, the API is built on the fact that usually less important arguments are placed at the end of the function signature and often they are optional parameters; mandatory parameters cannot be placed after optional ones in most programming languages. This assumption is incorrect for a few cases where you want to add some prefix or fix the index/key argument in methods with a signature like `set_something(index: int, value: Value)`, but that case doesn't seem to be very common compared to the opposite.

So I would suggest just documenting it and providing the snippet above as an example of implementing the "prebinding" functionality.

**Comment by AriaTheCurseling** [comment]
Yes, you can make a utility, but my primary reason for wanting prebing is so i can connect an arbitrary signal, say `value_changed(new_value` to a `set(property_name, value)` *through the editor*

I cannot use a utility function when connecting via the editor unless i create an instance of it *per connection i want to create*

Thats a massive clutter for relative simple functionality, and significantly limits the flexibility of signals.

**Comment by KoBeWi** [comment]
> Regarding adding the method out of the box, while I agree it could be useful, please keep in mind that it affects other APIs as well:

Does it? I just opened a PR, I only had to implement one method to make it work. `unbind()` works as you would expect, i.e. it trims the final argument list. `get_bound_arguments/count()` does not differentiate prebound arguments, but it does not really matter if the callable is called correctly.

**Comment by dalexeev** [comment]
I mean that you can unbind N arguments from both left and right. And this also affects the API for getting the bound/unbound argument info. If you can bind/unbind arguments from left and right, then it's ambiguous if you just return all arguments in one array, both left bound and right bound. If you want to fully implement this, you should add all the following methods:

* `Callable.get_prebound_arguments()`;
* `Callable.get_prebound_arguments_count()`;
* `Callable.get_preunbound_arguments_count()`;
* `Callable.preunbind()`.

**Comment by KoBeWi** [comment]
I don't think adding other methods for the sake of completeness is any useful. `get_*arguments()` are used mostly internally, `preubind()` does not even sound useful. And it's not like they can't be added later.

**Comment by dalexeev** [comment]
In my opinion, obvious asymmetry, incompleteness and inconsistency in the logical model usually indicate problems in the design. All other things being equal, I prefer a coherent logical model rather than taking into account only practical considerations. This is not a case where pragmatism is more important than purism.

Even if most regular users don't need it, the lack of this functionality can be very painful for plugin authors that use reflection and metaprogramming to some extent (test frameworks, debuggers, etc.).

Suppose `preunbind()` is not needed yet. Still, not being able to get prebound arguments looks wrong (inconsistent and asymmetric), and mixing prebound and postbound arguments in existing methods is wrong, as I pointed out above, you won't be able to distinguish them.

**Comment by tdbear** [comment]
Bind appending arguments from the right breaks the principle of least surprise for anyone who ever programmed in a language other than GDscript. Arguments in functions typically are ordered in decreasing "importance" or "effect" with the right most  arguments often having defaults or even variadics. Would you expect a typical use case for a bind to try to define the values for those parameters? No. Just saying.

**Comment by Lexyth** [comment]
> Arguments in functions typically are ordered in decreasing "importance" or "effect" with the right most arguments often having defaults or even variadics. Would you expect a typical use case for a bind to try to define the values for those parameters? No. Just saying.

I don't think the importance or effect of a parameter should play any role in the order they are bound in.
```gdscript
func doSomething (what, when = "now") ... # omitted.

... # omitted.

var doThat = doSomething.bind_at(0, that) # pseudo-code.
var doSomethingLater = doSomething.bind_at(1, "then") # pseudo-code.

doThat(when)
doSomethingThen(what)
```
Both are equally useful and thus equally likely. 
I would as much expect defaulted parameters or variadics to be bound as I would any other parameter.
